contracts/contracts/token/OUSD.sol

```diff
@@ -472,14 +499,18 @@ contract OUSD is Governable {
      */
     function _isNonRebasingAccount(address _account) internal returns (bool) {
         bool isContract = _account.code.length > 0;
-        if (isContract && rebaseState[_account] == RebaseOptions.NotSet && alternativeCreditsPerToken[_account] != 0) {
-            _rebaseOptOut(msg.sender);
+        if (isContract && rebaseState[_account] == RebaseOptions.NotSet) {
```

Comment:
🟥 !! `_rebaseOptOut` brutally changes an account's accounting (as well as making the resulting global update that follow). It is absolutely vital that it does not get called on anything that is already using a nonrebasing accounting. Stuff will go deeply wrong.

While the previous check in this function was wrong (it should be == 0), the check should not be removed. 

Basically, this code should only call `_rebaseOptOut()`, if the call to `_rebaseOptOut` is going to work, and `_rebaseOptOut` should check and revert on this.


We should also change the code back in `_rebaseOptOut()` first require:

```
require(alternativeCreditsPerToken[_account] == 0, "Account must be rebasing");
```

Unless there's something I'm missing?

Reply:
The code in `_rebaseOptOut()` has been changed to:

```
        if (state == RebaseOptions.StdRebasing) {
            require(
                alternativeCreditsPerToken[_account] == 0,
                "Account must be rebasing"
            );
        }
```

Almost all our current historical contracts do not have `RebaseOptions.StdRebasing` set on them. They just operate as NonSet. This means that by only doing this check, any contract that was created and had interacted with our tokens before is already nonRebasing, but could call `_rebaseOptOut()` a second time. This would be very, very bad, since it would throw off both the rebasingCredits and the nonRebasingSupply globals.

Reply:
It's probably worth reading <https://github.com/OriginProtocol/yolo-spike-ousd-token/tree/master> if you haven't already. Goes into what the old data looks like.

Reply:
Thanks for the explanation great points 🙏  

Some thoughts regarding the check in `isNonrebasingAccount` I do see that it was there from the underlying `ensureRebasingMigration` [function](https://github.com/OriginProtocol/origin-dollar/blob/master/contracts/contracts/token/OUSD.sol#L473). And we needed that check solely because the `ensureRebasingMigration` would "migrate the account" by changing global non rebasing supply and setting per account custom `nonRebasingCreditsPerToken` but it **would not** alter the rebase state. By having such check the migration part of the `ensureRebasingMigration` would not be executed twice for the same account. 

Now that we explicitly migrate an account by also altering `rebaseState` the additional check is not needed anymore for new accounts after the contract is deployed. But the state for the old accounts must still be handled and those might have existing non 0 `alternativeCreditsPerToken` with the rebase state set to `RebaseOptions.NotSet`. 

Added a comment here: https://github.com/OriginProtocol/origin-dollar/pull/2298/commits/47e529f0556577378f11c8f9d0812f2085ddcb88

Reply:
And yes you are right the check you had in `rebaseOptOut` completely makes sense. Added it back here: https://github.com/OriginProtocol/origin-dollar/pull/2298/commits/22c1cf5c8536b1b884fdcb4d815a2920581aa5f5